common/js/src/logging/crash-loop-detection.js

```diff
@@ -0,0 +1,132 @@
+// Chrome's Extensions system restrictions, which are "5 reloads, each within 10
+// seconds after the launch" (as of 2020-08-13).
+const CRASH_LOOP_THRESHOLD_COUNT = 4;
+
+/** @type {boolean} */
+let crashing = false;
+
+GSC.Logging.CrashLoopDetection.handleImminentCrash = function() {
+  if (crashing)
+    return Promise.reject('Already crashing');
+  crashing = true;
+
+  return loadRecentCrashTimestamps().then(recentCrashTimestamps => {
+    return storeNewCrashTimestamps(recentCrashTimestamps);
+  }).then(newCrashTimestamps => {
+    const isInCrashLoop =
+        newCrashTimestamps.length >= CRASH_LOOP_THRESHOLD_COUNT;
+    return Promise.resolve(isInCrashLoop);
+  });
+};
+
+/**
+ * Loads the recent crash timestamps from the persistent storage.
+ * @return {!Promise<!Array<number>>}
+ */
+function loadRecentCrashTimestamps() {
+  // Note that none of the error handling below throws assertions or emits
+  // logs, since this code is triggered from handling another, and more severe,
+  // error.
+  if (!chrome || !chrome.storage || !chrome.storage.local) {
+    // Cannot load any data without access to the chrome.storage API (most
+    // likely, due to a missing permission).
+    return Promise.resolve([]);
+  }
+  return new Promise(resolve => {
+    chrome.storage.local.get(STORAGE_KEY, function(loadedStorage) {
+      if (chrome.runtime.lastError ||
+          !loadedStorage.hasOwnProperty(STORAGE_KEY) ||
+          !Array.isArray(loadedStorage[STORAGE_KEY])) {
+        resolve([]);
+        return;
+      }
+      resolve(loadedStorage[STORAGE_KEY].filter(item => {
+        return Number.isInteger(item);
+      }));
+    });
+  });
+}
+
+/**
+ * Generates a new crash timestamps list (together with the current timestamp),
+ * stores it in the persistent storage and returns it.
+ * @param {!Array<number>} recentCrashTimestamps
+ * @return {!Promise<!Array<number>>}
+ */
+function storeNewCrashTimestamps(recentCrashTimestamps) {
+  const currentTimestamp = Date.now();
+  const minTimestampToKeep = currentTimestamp - CRASH_LOOP_WINDOW_MILLISECONDS;
+  let newCrashTimestamps = recentCrashTimestamps.filter(crashTimestamp => {
+    return minTimestampToKeep <= crashTimestamp &&
+           crashTimestamp <= currentTimestamp;
+  });
+  // Sanitization: Drop extra items beyond the |CRASH_LOOP_THRESHOLD_COUNT-1|
+  // ones, in order to avoid any risk of the storage growing beyond all bounds.
+  if (newCrashTimestamps.length >= CRASH_LOOP_THRESHOLD_COUNT) {
+    newCrashTimestamps = newCrashTimestamps.slice(
+        -CRASH_LOOP_THRESHOLD_COUNT + 1);
+  }
+  newCrashTimestamps.push(currentTimestamp);
+
+  if (!chrome || !chrome.storage || !chrome.storage.local) {
+    // Cannot store any data without access to the chrome.storage API.
+    return Promise.resolve(newCrashTimestamps);
+  }
+  return new Promise(resolve => {
+    chrome.storage.local.set({[STORAGE_KEY]: newCrashTimestamps}, function() {
```

Comment:
Question: Why is STORAGE_KEY inside [] here?

Reply:
It's the JS syntax (introduced in ECMAScript2015) that means "read the STORAGE_KEY's value", rather than "literally use the `"STORAGE_KEY"` string".

But I've replaced it with a separate object construction, so that there's no confusion.